### PR TITLE
Add support for incremental text synchronization

### LINKED ISSUE #3762: Incremental synchronization support
Spawned off #3758.

Some LSP clients (ahem, Gnome Builder) don't respect our choice for full text synchronization, and it wouldn't hurt to reduce the amount of text transferred.

The text change notification comes in [here](https://github.com/rust-analyzer/rust-analyzer/blob/1c2d4135db867efe335a0654d86429bea7bb9caf/crates/rust-analyzer/src/main_loop.rs#L598-L599). `params.content_changes` contains all [changes](https://docs.rs/lsp-types/0.73.0/lsp_types/struct.TextDocumentContentChangeEvent.html), but right now we only process one of them. Before passing the changes to the VFS layer, we need to convert from a line/column to byte offset representation. The latter is implemented as the `TextRange` type, and a conversion method is available [here](https://github.com/rust-analyzer/rust-analyzer/blob/1c2d4135db867efe335a0654d86429bea7bb9caf/crates/rust-analyzer/src/conv.rs#L205).

### PR DESCRIPTION
**Pull request by lnicola**
Fixes #3762.

This still needs a `ra_vfs` PR, but I want to know I'm on the right track. I tested the change and it didn't crash horribly, but YMMV.

### CODE REVIEW

File: crates/rust-analyzer/src/main_loop.rs [hunk 1]
```diff
@@ -579,12 +582,24 @@ fn on_notification(
...
             let path = uri.to_file_path().map_err(|()| format!("invalid uri: {}", uri))?;
-            let text =
-                params.content_changes.pop().ok_or_else(|| "empty changes".to_string())?.text;
-            state.vfs.write().change_file_overlay(path.as_path(), text);
+            state.vfs.write().change_file_overlay(path.as_path(), |old_text| {
+                for change in content_changes {
+                    match change.range {
+                        Some(range) => {
+                            let range = range.conv_with(&line_index);
+                            old_text.replace_range(Range::<usize>::from(range), &change.text);
```
Comment (lnicola) [inline crates/rust-analyzer/src/main_loop.rs:597]: Should `TextRange` impl `RangeBounds<usize>`?

**Comment by kjeremy** [comment]
As an aside I'm surprised that this isn't something negotiated by the LSP using client/server caps. It seems weird to me that a client must support all three modes: None, Full and Incremental.

**Comment by kjeremy** [comment]
@Inicola does this improve responsiveness? I would expect the over the wire performance to be negligible but it looks like we avoid a 200ms delay in the vscode client in incremental mode: https://github.com/microsoft/vscode-languageserver-node/blob/e5d7ad881b1a51b486e3f0e4aa0fbc25dad2be58/client/src/client.ts#L1004

**Comment by lnicola** [comment]
I can't really tell, I only checked that it works. The type hints seemed pretty snappy, though. But when looking into the Builder failure I tried to sniff the client-server communication and couldn't notice the delay while typing.

I hope this won't make RA burn more CPU than before.

**Comment by kjeremy** [comment]
I just did a quick smoke test of this and it seems to work. I didn't spot any regressions.

**Comment by lnicola** [comment]
Does it feel faster?

**Comment by kjeremy** [comment]
The highlighting felt snappier.

**Comment by matklad** [comment]
This seems reasonable, though I am not sure if it works correctly with our `\r\n` translation. 

Let's add some unit-tests here, which create an empty VFS and apply some hard-coded JSONs to it?

**Comment by matklad** [comment]
(the vfs PR is published)

**Comment by lnicola** [comment]
I'm not sure how to structure it, is there another test working with JSON requests like that? Maybe we should add one to `ra_vfs`.

**Comment by matklad** [comment]
I mean, the tricky bit here is correctly interpreting `DidChangeTextDocumentParams` json (specifically, that we correctly transition positions, and we correclty handle \r\n)

I think just extracting [this bit](https://github.com/rust-analyzer/rust-analyzer/pull/4153/files#diff-688b34883e8efae680c3c682383af7e3R592-R602) into a function and unit-testing it should be possible. The funciton would take two arguments: VFS and a line index, and it would apply changes to VFS. 

The test would take initial file contents, the `DidChangeTextDocumentParams`, and the expected file context. it'll crate a VFS with a single file, calculate line index for it, apply the extracted function and compare results. 

I think thre required `DidChangeTextDocumentParams` can be copy-pasted from the lsp log messages? using multiple cursor and doing multi-line edits should cover enough interesting cases.

**Comment by lnicola** [comment]
I'm so confused -- I can't really figure out how to use `Vfs`. I can create one, but do I need to provide a real (disk) path, or can I just call `Vfs::load`? And then I must call its `handle_task` method with the event it returns? Finally, how do I read a file? `VfsFileData` is private.

**Comment by matklad** [comment]
Hm, indeed, I think because the scanning happens in background it's not really possible to test this correctly with VFS :-( 

Here's as far as I got:

```rust
#[test]
fn smoked_vfs() {
    struct All;
    impl ra_vfs::Filter for All {
        fn include_dir(&self, dir_path: &relative_path::RelativePath) -> bool {
            true
        }
        fn include_file(&self, file_path: &relative_path::RelativePath) -> bool {
            true
        }
    }
    let root = RootEntry::new("/dummy".into(), Box::new(All));
    let (mut vfs, vfs_roots) =
        Vfs::new(vec![root], Box::new(|task| drop(task) /* :-( */), Watch(false));

    let file_id = vfs.add_file_overlay(Path::new("/dummy/foo.rs"), "fn main() {}".into()).unwrap();
    let changes = vfs.commit_changes();
    match changes.as_slice() {
        [VfsChange::AddFile { text, .. }] => text.as_str() == "fn main() {}",
        changes => panic!("{:?}", changes),
    };

    vfs.change_file_overlay(Path::new("/dummy/foo.rs"), "fn bar() {}".into());
    vfs.commit_changes();
    match changes.as_slice() {
        [VfsChange::ChangeFile { text, .. }] => text.as_str() == "fn main() {}",
        changes => panic!("{:?}", changes),
    };
}
```

But perhaps we don't even need VFS at all? it seems like

```rust
fn apply_change(chage: lsp_types::TextDocumentChange, original_text: &mut String, line_endings: LineEndings)
```

Should actually be enough there?

What I am trying to test here is that we correctly interpret `TextDocumentChange`. Hm, actually, I think we might not be doing that right now? IIRC, it has somewhat insane API in that each atomic change should be applied in order. That is, the "coordinate space" of each atomic change depends on preceeding changes. But that means that we might actually need to recompute/adjust `LineIndex` appropriately?

**Comment by lnicola** [comment]
I've been worried about this from the start -- I'm not sure how we're supposed to handle multiple changed, maybe we should take a look at other language servers.

One thing I noticed is that Code tries to sort the ranges in reverse, so they don't conflict. I didn't test inserting multiple newlines, but it might work.

**Comment by lnicola** [comment]
https://github.com/Microsoft/language-server-protocol/pull/289 applying the changes in order seems to be correct.

**Comment by matklad** [comment]
>applying the changes in order seems to be correct.

Yes, they need to be applied in order, but the question is, will the offsets shift? The wont', if the changes are indeed reveres-sorted, but that's not a part of the protocol :-(

**Comment by kjeremy** [comment]
I think that's a question for @dbaeumer.

**Comment by lnicola** [comment]
Just to make sure I got it right: if a file changes, we invalidate the line index at the turn of the main loop, so we're worried about it becoming stale while we're applying the updates? And if the ranges are sorted backwards, it's not an issue.

**Comment by matklad** [comment]
Right. Specifically, let's say we start with an empty file, and the frist added adds three `\n` characters, and the second edit adds `"hello"` at the end. For the second edit, the line will be `3`. But the line index we have is empty, so we'll panic when resolving that `line three` to numerical offset. 

The correct (albeit inneficient) algorithm would be:

```
let mut text = initial_text;
for change in changes {
    let line_index = LineIndex::new(&text);
    let text_edit = change.conv_with(&line_index);
    text = text_edit.apply(text);
}
```

**Comment by lnicola** [comment]
I checked a couple of implementations:

 - Go doesn't use a line index: https://github.com/sourcegraph/go-langserver/blob/master/langserver/ast.go#L15
 - TypeScript doesn't support incremental synchronization: https://github.com/sourcegraph/javascript-typescript-langserver/blob/master/src/typescript-service.ts#L1614
 - clangd doesn't use an index: https://reviews.llvm.org/diffusion/L/browse/clang-tools-extra/trunk/clangd/SourceCode.cpp$149
 - the Dart analysis server rebuilds the index on each change: https://github.com/dart-lang/sdk/blob/master/pkg/analysis_server/lib/src/lsp/source_edits.dart#L11

File: crates/rust-analyzer/src/main_loop.rs [hunk 2]
```diff
@@ -653,6 +662,44 @@ fn on_notification(
...
+            }
+        }
+    }
+
+    let mut index_valid = IndexValid::All;
+    for change in content_changes {
+        match change.range {
+            Some(range) => {
+                if !index_valid.covers(range.start.line) {
+                    line_index = Cow::Owned(LineIndex::new(&old_text));
```
Comment (matklad) [inline crates/rust-analyzer/src/main_loop.rs:689]: :exploding_head: that's ingenious!

Comment (matklad) [reply crates/rust-analyzer/src/main_loop.rs:689]: still needs tests tho ;)

Comment (lnicola) [reply crates/rust-analyzer/src/main_loop.rs:693]: I added a couple, can you take a look?

**Comment by matklad** [comment]
bors d+

We *can* add a temporary flag to config, to allow users to opt-out if this misbehaves (and I think this can misbehave both due to our bugs, and due to client bugs). OTOH, I am also ok with merging this as is, so, at your discretion.

**Comment by lnicola** [comment]
The capabilities [seem to be](https://github.com/rust-analyzer/rust-analyzer/blob/master/crates/rust-analyzer/src/bin/main.rs#L77-L79) sent right from the start, so I'm not sure how we can disable incremental sync by configuration.

Let's merge it and we can disable it before Monday if anyone complains.

bors r=matklad

**Comment by kjeremy** [comment]
The capabilities thing is a bug in lsp-server (see https://github.com/rust-analyzer/lsp-server/pull/14) so I think we could look at the initial params or experimental client caps and change the mode if we need to.

File: crates/rust-analyzer/src/main_loop.rs (hunk 1 above)
Comment (lnicola) [reply crates/rust-analyzer/src/main_loop.rs]: It can't, it needs to return a reference.

**Comment by dbaeumer** [comment]
Do you still need an answer on this: https://github.com/rust-analyzer/rust-analyzer/pull/4153#issuecomment-621747110

```
	/**
	 * The actual content changes. The content changes describe single state changes
	 * to the document. So if there are two content changes c1 (at array index 0) and
	 * c2 (at array index 1) for a document in state S then c1 moves the document from
	 * S to S' and c2 from S' to S''. So c1 is computed on the state S and c2 is computed
	 * on the state S'.
	 *
	 * To mirror the content of a document using change events use the following approach:
	 * - start with the same initial content
	 * - apply the 'textDocument/didChange' notifications in the order you recevie them.
	 * - apply the `TextDocumentContentChangeEvent`s in a single notification in the order
	 *   you receive them.
	 */
```

Should explain everything.
